code/modules/recycling/disposal/holder.dm

```diff
@@ -0,0 +1,130 @@
+
+/obj/structure/disposalholder/Destroy()
+	qdel(gas)
+	active = FALSE
+	return ..()
+
+	// initialize a holder from the contents of a disposal unit
+/obj/structure/disposalholder/proc/init(obj/machinery/disposal/D)
+	gas = D.air_contents// transfer gas resv. into holder object
+
+	//Check for any living mobs trigger hasmob.
+	//hasmob effects whether the package goes to cargo or its tagged destination.
+	for(var/mob/living/M in D)
+		if(M.client)
+			M.reset_perspective(src)
+		hasmob = TRUE
+
+	//Checks 1 contents level deep. This means that players can be sent through disposals mail...
+	//...but it should require a second person to open the package. (i.e. person inside a wrapped locker)
+	for(var/obj/O in D)
+		if(O.contents)
+			for(var/mob/living/M in O.contents)
+				hasmob = TRUE
+
+	// now everything inside the disposal gets put into the holder
+	// note AM since can contain mobs or objs
+	for(var/atom/movable/AM in D)
+		AM.loc = src
+		if(istype(AM, /obj/structure/bigDelivery) && !hasmob)
+			var/obj/structure/bigDelivery/T = AM
+			src.destinationTag = T.sortTag
+		if(istype(AM, /obj/item/smallDelivery) && !hasmob)
+			var/obj/item/smallDelivery/T = AM
+			src.destinationTag = T.sortTag
+
+
+// start the movement process
+// argument is the disposal unit the holder started in
+/obj/structure/disposalholder/proc/start(obj/machinery/disposal/D)
+	if(!D.trunk)
+		D.expel(src)	// no trunk connected, so expel immediately
+		return
+	loc = D.trunk
+	active = TRUE
+	setDir(DOWN)
+	move()
+
+	return
+
+// movement process, persists while holder is moving through pipes
+/obj/structure/disposalholder/proc/move()
+	set waitfor = 0
+	var/obj/structure/disposalpipe/last
+	while(active)
+		var/obj/structure/disposalpipe/curr = loc
+		last = curr
+		curr = curr.transfer(src)
+		if(!curr && active)
+			last.expel(src, loc, dir)
+
+		stoplag()
+		if(!(count--))
+			active = FALSE
+	return
+
+// find the turf which should contain the next pipe
+/obj/structure/disposalholder/proc/nextloc()
+	return get_step(loc,dir)
+
+// find a matching pipe on a turf
+/obj/structure/disposalholder/proc/findpipe(turf/T)
+	if(!T)
+		return null
+
+	var/fdir = turn(dir, 180)	// flip the movement direction
+	for(var/obj/structure/disposalpipe/P in T)
+		if(fdir & P.dpdir)		// find pipe direction mask that matches flipped dir
+			return P
+	// if no matching pipe, return null
+	return null
+
+// merge two holder objects
+// used when a holder meets a stuck holder
+/obj/structure/disposalholder/proc/merge(obj/structure/disposalholder/other)
+	for(var/atom/movable/AM in other)
+		AM.loc = src		// move everything in other holder to this one
+		if(ismob(AM))
+			var/mob/M = AM
+			M.reset_perspective(src)	// if a client mob, update eye to follow this holder
+	qdel(other)
+
+
+// called when player tries to move while in a pipe
+/obj/structure/disposalholder/relaymove(mob/user)
+	if (user.stat)
+		return
+	if (src.loc)
+		for (var/mob/M in get_hearers_in_view(src.loc.loc))
```

Comment:
If you want the turf, use get_turf(src). If you want the area, use get_area(src)